Index: src/com/hp/hpl/jena/graph/impl/LiteralLabel.java
===================================================================
RCS file: /cvsroot/jena/jena2/src/com/hp/hpl/jena/graph/impl/Attic/LiteralLabel.java,v
retrieving revision 1.39
diff -u -r1.39 LiteralLabel.java
--- src/com/hp/hpl/jena/graph/impl/LiteralLabel.java	27 Mar 2009 15:27:51 -0000	1.39
+++ src/com/hp/hpl/jena/graph/impl/LiteralLabel.java	25 Mar 2010 15:17:08 -0000
@@ -63,6 +63,12 @@
 	 */
 	private boolean wellformed = true;
 
+	/**
+	* keeps the message provided by the DatatypeFormatException
+	* if parsing failed for delayed exception thrown in getValue()
+	*/
+	private String exceptionMsg = null;
+	
 	//=======================================================================
 	// Constructors
 
@@ -199,6 +205,7 @@
 				throw e;
 			} else {
 				wellformed = false;
+				exceptionMsg = e.getMessage();
 			}
 		}
 	}
@@ -293,7 +300,7 @@
 			throw new DatatypeFormatException(
 				lexicalForm,
 				dtype,
-				" in getValue()");
+				exceptionMsg);
 		}
 	}
 
Index: src/com/hp/hpl/jena/graph/Node.java
===================================================================
RCS file: /cvsroot/jena/jena2/src/com/hp/hpl/jena/graph/Attic/Node.java,v
retrieving revision 1.62
diff -u -r1.62 Node.java
--- src/com/hp/hpl/jena/graph/Node.java	16 Jan 2009 17:23:52 -0000	1.62
+++ src/com/hp/hpl/jena/graph/Node.java	25 Mar 2010 15:17:08 -0000
@@ -24,7 +24,9 @@
 */
 
 public abstract class Node {
-    
+	// AndyL provenance hack
+	private String source;
+	
     final protected Object label;
     static final int THRESHOLD = 10000;
     
@@ -330,6 +332,14 @@
     public boolean matches( Node other )
         { return equals( other ); }
 
+    public void setSource(String source) {
+    	this.source = source;
+    }
+    
+    public String getSource() {
+    	return source;
+    }
+    
     /** 
         Answer a human-readable representation of this Node. It will not compress URIs, 
         nor quote literals (because at the moment too many places use toString() for 
