Index: src/de/fuberlin/wiwiss/ng4j/trig/parser/TriGAntlrLexer.java
===================================================================
RCS file: /cvsroot/ng4j/ng4j/src/de/fuberlin/wiwiss/ng4j/trig/parser/TriGAntlrLexer.java,v
retrieving revision 1.5
diff -u -r1.5 TriGAntlrLexer.java
--- src/de/fuberlin/wiwiss/ng4j/trig/parser/TriGAntlrLexer.java	15 Jun 2007 16:05:31 -0000	1.5
+++ src/de/fuberlin/wiwiss/ng4j/trig/parser/TriGAntlrLexer.java	25 Mar 2010 15:44:32 -0000
@@ -223,10 +223,10 @@
 		_ttype = THING;
 		int _saveIndex;
 		
-		boolean synPredMatched65 = false;
+		boolean synPredMatched66 = false;
 		if (((LA(1)=='h') && (LA(2)=='a') && (LA(3)=='s'))) {
-			int _m65 = mark();
-			synPredMatched65 = true;
+			int _m66 = mark();
+			synPredMatched66 = true;
 			inputState.guessing++;
 			try {
 				{
@@ -235,22 +235,22 @@
 				}
 			}
 			catch (RecognitionException pe) {
-				synPredMatched65 = false;
+				synPredMatched66 = false;
 			}
-			rewind(_m65);
+			rewind(_m66);
 			inputState.guessing--;
 		}
-		if ( synPredMatched65 ) {
+		if ( synPredMatched66 ) {
 			match("has");
 			if ( inputState.guessing==0 ) {
 				_ttype = KW_HAS ;
 			}
 		}
 		else {
-			boolean synPredMatched69 = false;
+			boolean synPredMatched70 = false;
 			if (((LA(1)=='t') && (LA(2)=='h') && (LA(3)=='i'))) {
-				int _m69 = mark();
-				synPredMatched69 = true;
+				int _m70 = mark();
+				synPredMatched70 = true;
 				inputState.guessing++;
 				try {
 					{
@@ -259,22 +259,22 @@
 					}
 				}
 				catch (RecognitionException pe) {
-					synPredMatched69 = false;
+					synPredMatched70 = false;
 				}
-				rewind(_m69);
+				rewind(_m70);
 				inputState.guessing--;
 			}
-			if ( synPredMatched69 ) {
+			if ( synPredMatched70 ) {
 				match("this");
 				if ( inputState.guessing==0 ) {
 					_ttype = KW_THIS ;
 				}
 			}
 			else {
-				boolean synPredMatched61 = false;
+				boolean synPredMatched62 = false;
 				if (((LA(1)==':') && (LA(2)=='-') && (true))) {
-					int _m61 = mark();
-					synPredMatched61 = true;
+					int _m62 = mark();
+					synPredMatched62 = true;
 					inputState.guessing++;
 					try {
 						{
@@ -283,22 +283,22 @@
 						}
 					}
 					catch (RecognitionException pe) {
-						synPredMatched61 = false;
+						synPredMatched62 = false;
 					}
-					rewind(_m61);
+					rewind(_m62);
 					inputState.guessing--;
 				}
-				if ( synPredMatched61 ) {
+				if ( synPredMatched62 ) {
 					match(":-");
 					if ( inputState.guessing==0 ) {
 						_ttype = NAME_OP ;
 					}
 				}
 				else {
-					boolean synPredMatched67 = false;
+					boolean synPredMatched68 = false;
 					if (((LA(1)=='o') && (LA(2)=='f') && (true))) {
-						int _m67 = mark();
-						synPredMatched67 = true;
+						int _m68 = mark();
+						synPredMatched68 = true;
 						inputState.guessing++;
 						try {
 							{
@@ -307,22 +307,22 @@
 							}
 						}
 						catch (RecognitionException pe) {
-							synPredMatched67 = false;
+							synPredMatched68 = false;
 						}
-						rewind(_m67);
+						rewind(_m68);
 						inputState.guessing--;
 					}
-					if ( synPredMatched67 ) {
+					if ( synPredMatched68 ) {
 						match("of");
 						if ( inputState.guessing==0 ) {
 							_ttype = KW_OF ;
 						}
 					}
 					else {
-						boolean synPredMatched73 = false;
+						boolean synPredMatched74 = false;
 						if (((LA(1)=='i') && (LA(2)=='s') && (true))) {
-							int _m73 = mark();
-							synPredMatched73 = true;
+							int _m74 = mark();
+							synPredMatched74 = true;
 							inputState.guessing++;
 							try {
 								{
@@ -331,22 +331,22 @@
 								}
 							}
 							catch (RecognitionException pe) {
-								synPredMatched73 = false;
+								synPredMatched74 = false;
 							}
-							rewind(_m73);
+							rewind(_m74);
 							inputState.guessing--;
 						}
-						if ( synPredMatched73 ) {
+						if ( synPredMatched74 ) {
 							match("is");
 							if ( inputState.guessing==0 ) {
 								_ttype = KW_IS ;
 							}
 						}
 						else {
-							boolean synPredMatched53 = false;
+							boolean synPredMatched54 = false;
 							if (((_tokenSet_3.member(LA(1))) && (true) && (true))) {
-								int _m53 = mark();
-								synPredMatched53 = true;
+								int _m54 = mark();
+								synPredMatched54 = true;
 								inputState.guessing++;
 								try {
 									{
@@ -356,12 +356,12 @@
 									}
 								}
 								catch (RecognitionException pe) {
-									synPredMatched53 = false;
+									synPredMatched54 = false;
 								}
-								rewind(_m53);
+								rewind(_m54);
 								inputState.guessing--;
 							}
-							if ( synPredMatched53 ) {
+							if ( synPredMatched54 ) {
 								mNSNAME(false);
 								mCOLON(false);
 								mLNAME(false);
@@ -370,10 +370,10 @@
 								}
 							}
 							else {
-								boolean synPredMatched55 = false;
+								boolean synPredMatched56 = false;
 								if (((LA(1)==':') && (true) && (true))) {
-									int _m55 = mark();
-									synPredMatched55 = true;
+									int _m56 = mark();
+									synPredMatched56 = true;
 									inputState.guessing++;
 									try {
 										{
@@ -382,12 +382,12 @@
 										}
 									}
 									catch (RecognitionException pe) {
-										synPredMatched55 = false;
+										synPredMatched56 = false;
 									}
-									rewind(_m55);
+									rewind(_m56);
 									inputState.guessing--;
 								}
-								if ( synPredMatched55 ) {
+								if ( synPredMatched56 ) {
 									mCOLON(false);
 									mLNAME(false);
 									if ( inputState.guessing==0 ) {
@@ -395,10 +395,10 @@
 									}
 								}
 								else {
-									boolean synPredMatched57 = false;
+									boolean synPredMatched58 = false;
 									if (((_tokenSet_3.member(LA(1))) && (true) && (true))) {
-										int _m57 = mark();
-										synPredMatched57 = true;
+										int _m58 = mark();
+										synPredMatched58 = true;
 										inputState.guessing++;
 										try {
 											{
@@ -407,12 +407,12 @@
 											}
 										}
 										catch (RecognitionException pe) {
-											synPredMatched57 = false;
+											synPredMatched58 = false;
 										}
-										rewind(_m57);
+										rewind(_m58);
 										inputState.guessing--;
 									}
-									if ( synPredMatched57 ) {
+									if ( synPredMatched58 ) {
 										mNSNAME(false);
 										mCOLON(false);
 										if ( inputState.guessing==0 ) {
@@ -420,10 +420,10 @@
 										}
 									}
 									else {
-										boolean synPredMatched59 = false;
+										boolean synPredMatched60 = false;
 										if (((LA(1)==':') && (true) && (true))) {
-											int _m59 = mark();
-											synPredMatched59 = true;
+											int _m60 = mark();
+											synPredMatched60 = true;
 											inputState.guessing++;
 											try {
 												{
@@ -431,22 +431,22 @@
 												}
 											}
 											catch (RecognitionException pe) {
-												synPredMatched59 = false;
+												synPredMatched60 = false;
 											}
-											rewind(_m59);
+											rewind(_m60);
 											inputState.guessing--;
 										}
-										if ( synPredMatched59 ) {
+										if ( synPredMatched60 ) {
 											mCOLON(false);
 											if ( inputState.guessing==0 ) {
 												_ttype = QNAME ;
 											}
 										}
 										else {
-											boolean synPredMatched63 = false;
+											boolean synPredMatched64 = false;
 											if (((_tokenSet_4.member(LA(1))) && (true) && (true))) {
-												int _m63 = mark();
-												synPredMatched63 = true;
+												int _m64 = mark();
+												synPredMatched64 = true;
 												inputState.guessing++;
 												try {
 													{
@@ -454,22 +454,22 @@
 													}
 												}
 												catch (RecognitionException pe) {
-													synPredMatched63 = false;
+													synPredMatched64 = false;
 												}
-												rewind(_m63);
+												rewind(_m64);
 												inputState.guessing--;
 											}
-											if ( synPredMatched63 ) {
+											if ( synPredMatched64 ) {
 												mNUMBER(false);
 												if ( inputState.guessing==0 ) {
 													_ttype = NUMBER ;
 												}
 											}
 											else {
-												boolean synPredMatched71 = false;
+												boolean synPredMatched72 = false;
 												if (((LA(1)=='a') && (true) && (true))) {
-													int _m71 = mark();
-													synPredMatched71 = true;
+													int _m72 = mark();
+													synPredMatched72 = true;
 													inputState.guessing++;
 													try {
 														{
@@ -478,12 +478,12 @@
 														}
 													}
 													catch (RecognitionException pe) {
-														synPredMatched71 = false;
+														synPredMatched72 = false;
 													}
-													rewind(_m71);
+													rewind(_m72);
 													inputState.guessing--;
 												}
-												if ( synPredMatched71 ) {
+												if ( synPredMatched72 ) {
 													match("a");
 													if ( inputState.guessing==0 ) {
 														_ttype = KW_A ;
@@ -569,24 +569,24 @@
 		}
 		}
 		{
-		int _cnt121=0;
-		_loop121:
+		int _cnt122=0;
+		_loop122:
 		do {
 			if (((LA(1) >= '0' && LA(1) <= '9'))) {
 				matchRange('0','9');
 			}
 			else {
-				if ( _cnt121>=1 ) { break _loop121; } else {throw new NoViableAltForCharException((char)LA(1), getFilename(), getLine(), getColumn());}
+				if ( _cnt122>=1 ) { break _loop122; } else {throw new NoViableAltForCharException((char)LA(1), getFilename(), getLine(), getColumn());}
 			}
 			
-			_cnt121++;
+			_cnt122++;
 		} while (true);
 		}
 		{
-		boolean synPredMatched125 = false;
+		boolean synPredMatched126 = false;
 		if (((LA(1)=='.'))) {
-			int _m125 = mark();
-			synPredMatched125 = true;
+			int _m126 = mark();
+			synPredMatched126 = true;
 			inputState.guessing++;
 			try {
 				{
@@ -597,25 +597,25 @@
 				}
 			}
 			catch (RecognitionException pe) {
-				synPredMatched125 = false;
+				synPredMatched126 = false;
 			}
-			rewind(_m125);
+			rewind(_m126);
 			inputState.guessing--;
 		}
-		if ( synPredMatched125 ) {
+		if ( synPredMatched126 ) {
 			mDOT(false);
 			{
-			int _cnt127=0;
-			_loop127:
+			int _cnt128=0;
+			_loop128:
 			do {
 				if (((LA(1) >= '0' && LA(1) <= '9'))) {
 					matchRange('0','9');
 				}
 				else {
-					if ( _cnt127>=1 ) { break _loop127; } else {throw new NoViableAltForCharException((char)LA(1), getFilename(), getLine(), getColumn());}
+					if ( _cnt128>=1 ) { break _loop128; } else {throw new NoViableAltForCharException((char)LA(1), getFilename(), getLine(), getColumn());}
 				}
 				
-				_cnt127++;
+				_cnt128++;
 			} while (true);
 			}
 		}
@@ -668,17 +668,17 @@
 			}
 			}
 			{
-			int _cnt132=0;
-			_loop132:
+			int _cnt133=0;
+			_loop133:
 			do {
 				if (((LA(1) >= '0' && LA(1) <= '9'))) {
 					matchRange('0','9');
 				}
 				else {
-					if ( _cnt132>=1 ) { break _loop132; } else {throw new NoViableAltForCharException((char)LA(1), getFilename(), getLine(), getColumn());}
+					if ( _cnt133>=1 ) { break _loop133; } else {throw new NoViableAltForCharException((char)LA(1), getFilename(), getLine(), getColumn());}
 				}
 				
-				_cnt132++;
+				_cnt133++;
 			} while (true);
 			}
 		}
@@ -713,10 +713,10 @@
 		_ttype = URI_OR_IMPLIES;
 		int _saveIndex;
 		
-		boolean synPredMatched78 = false;
+		boolean synPredMatched79 = false;
 		if (((LA(1)=='<') && (LA(2)=='=') && (LA(3)=='>'))) {
-			int _m78 = mark();
-			synPredMatched78 = true;
+			int _m79 = mark();
+			synPredMatched79 = true;
 			inputState.guessing++;
 			try {
 				{
@@ -724,22 +724,22 @@
 				}
 			}
 			catch (RecognitionException pe) {
-				synPredMatched78 = false;
+				synPredMatched79 = false;
 			}
-			rewind(_m78);
+			rewind(_m79);
 			inputState.guessing--;
 		}
-		if ( synPredMatched78 ) {
+		if ( synPredMatched79 ) {
 			mARROW_MEANS(false);
 			if ( inputState.guessing==0 ) {
 				_ttype = ARROW_MEANS ;
 			}
 		}
 		else {
-			boolean synPredMatched76 = false;
+			boolean synPredMatched77 = false;
 			if (((LA(1)=='<') && (LA(2)=='=') && (true))) {
-				int _m76 = mark();
-				synPredMatched76 = true;
+				int _m77 = mark();
+				synPredMatched77 = true;
 				inputState.guessing++;
 				try {
 					{
@@ -747,12 +747,12 @@
 					}
 				}
 				catch (RecognitionException pe) {
-					synPredMatched76 = false;
+					synPredMatched77 = false;
 				}
-				rewind(_m76);
+				rewind(_m77);
 				inputState.guessing--;
 			}
-			if ( synPredMatched76 ) {
+			if ( synPredMatched77 ) {
 				mARROW_L(false);
 				if ( inputState.guessing==0 ) {
 					_ttype = ARROW_L ;
@@ -810,17 +810,17 @@
 		mLANGLE(false);
 		text.setLength(_saveIndex);
 		{
-		_loop82:
+		_loop83:
 		do {
 			// nongreedy exit test
-			if ((LA(1)=='>') && (true)) break _loop82;
+			if ((LA(1)=='>') && (true)) break _loop83;
 			if ((_tokenSet_0.member(LA(1))) && (_tokenSet_0.member(LA(2)))) {
 				{
 				match(_tokenSet_0);
 				}
 			}
 			else {
-				break _loop82;
+				break _loop83;
 			}
 			
 		} while (true);
@@ -1105,17 +1105,17 @@
 		
 		mQUESTION(false);
 		{
-		int _cnt86=0;
-		_loop86:
+		int _cnt87=0;
+		_loop87:
 		do {
 			if ((_tokenSet_6.member(LA(1)))) {
 				mALPHANUMERIC(false);
 			}
 			else {
-				if ( _cnt86>=1 ) { break _loop86; } else {throw new NoViableAltForCharException((char)LA(1), getFilename(), getLine(), getColumn());}
+				if ( _cnt87>=1 ) { break _loop87; } else {throw new NoViableAltForCharException((char)LA(1), getFilename(), getLine(), getColumn());}
 			}
 			
-			_cnt86++;
+			_cnt87++;
 		} while (true);
 		}
 		if ( _createToken && _token==null && _ttype!=Token.SKIP ) {
@@ -1144,10 +1144,10 @@
 		int _saveIndex;
 		Token a=null;
 		
-		boolean synPredMatched89 = false;
+		boolean synPredMatched90 = false;
 		if (((LA(1)=='@') && (LA(2)=='p') && (LA(3)=='r'))) {
-			int _m89 = mark();
-			synPredMatched89 = true;
+			int _m90 = mark();
+			synPredMatched90 = true;
 			inputState.guessing++;
 			try {
 				{
@@ -1156,12 +1156,12 @@
 				}
 			}
 			catch (RecognitionException pe) {
-				synPredMatched89 = false;
+				synPredMatched90 = false;
 			}
-			rewind(_m89);
+			rewind(_m90);
 			inputState.guessing--;
 		}
-		if ( synPredMatched89 ) {
+		if ( synPredMatched90 ) {
 			mAT(false);
 			match("prefix");
 			if ( inputState.guessing==0 ) {
@@ -1169,10 +1169,10 @@
 			}
 		}
 		else {
-			boolean synPredMatched92 = false;
+			boolean synPredMatched93 = false;
 			if (((LA(1)=='@') && (_tokenSet_1.member(LA(2))) && (true))) {
-				int _m92 = mark();
-				synPredMatched92 = true;
+				int _m93 = mark();
+				synPredMatched93 = true;
 				inputState.guessing++;
 				try {
 					{
@@ -1183,12 +1183,12 @@
 					}
 				}
 				catch (RecognitionException pe) {
-					synPredMatched92 = false;
+					synPredMatched93 = false;
 				}
-				rewind(_m92);
+				rewind(_m93);
 				inputState.guessing--;
 			}
-			if ( synPredMatched92 ) {
+			if ( synPredMatched93 ) {
 				mAT(false);
 				{
 				int _cnt_a=0;
@@ -1208,13 +1208,13 @@
 				if ((LA(1)=='-')) {
 					match("-");
 					{
-					_loop97:
+					_loop98:
 					do {
 						if ((_tokenSet_1.member(LA(1)))) {
 							mALPHA(false);
 						}
 						else {
-							break _loop97;
+							break _loop98;
 						}
 						
 					} while (true);
@@ -1460,13 +1460,13 @@
 		int _saveIndex;
 		
 		{
-		_loop115:
+		_loop116:
 		do {
 			if ((_tokenSet_7.member(LA(1)))) {
 				mXNAMECHAR(false);
 			}
 			else {
-				break _loop115;
+				break _loop116;
 			}
 			
 		} while (true);
@@ -1526,10 +1526,10 @@
 		_ttype = STRING1;
 		int _saveIndex;
 		
-		boolean synPredMatched193 = false;
+		boolean synPredMatched194 = false;
 		if (((LA(1)=='\'') && (LA(2)=='\'') && (LA(3)=='\''))) {
-			int _m193 = mark();
-			synPredMatched193 = true;
+			int _m194 = mark();
+			synPredMatched194 = true;
 			inputState.guessing++;
 			try {
 				{
@@ -1537,24 +1537,24 @@
 				}
 			}
 			catch (RecognitionException pe) {
-				synPredMatched193 = false;
+				synPredMatched194 = false;
 			}
-			rewind(_m193);
+			rewind(_m194);
 			inputState.guessing--;
 		}
-		if ( synPredMatched193 ) {
+		if ( synPredMatched194 ) {
 			_saveIndex=text.length();
 			mQUOTE3S(false);
 			text.setLength(_saveIndex);
 			{
-			_loop198:
+			_loop199:
 			do {
 				// nongreedy exit test
-				if ((LA(1)=='\'') && (LA(2)=='\'') && (LA(3)=='\'')) break _loop198;
-				boolean synPredMatched196 = false;
+				if ((LA(1)=='\'') && (LA(2)=='\'') && (LA(3)=='\'')) break _loop199;
+				boolean synPredMatched197 = false;
 				if (((LA(1)=='\n'||LA(1)=='\r') && ((LA(2) >= '\u0000' && LA(2) <= '\ufffe')) && ((LA(3) >= '\u0000' && LA(3) <= '\ufffe')))) {
-					int _m196 = mark();
-					synPredMatched196 = true;
+					int _m197 = mark();
+					synPredMatched197 = true;
 					inputState.guessing++;
 					try {
 						{
@@ -1562,12 +1562,12 @@
 						}
 					}
 					catch (RecognitionException pe) {
-						synPredMatched196 = false;
+						synPredMatched197 = false;
 					}
-					rewind(_m196);
+					rewind(_m197);
 					inputState.guessing--;
 				}
-				if ( synPredMatched196 ) {
+				if ( synPredMatched197 ) {
 					mNL(false);
 				}
 				else if ((_tokenSet_8.member(LA(1))) && ((LA(2) >= '\u0000' && LA(2) <= '\ufffe')) && ((LA(3) >= '\u0000' && LA(3) <= '\ufffe'))) {
@@ -1579,7 +1579,7 @@
 					mESCAPE(false);
 				}
 				else {
-					break _loop198;
+					break _loop199;
 				}
 				
 			} while (true);
@@ -1593,10 +1593,10 @@
 			match('\'');
 			text.setLength(_saveIndex);
 			{
-			_loop200:
+			_loop201:
 			do {
 				// nongreedy exit test
-				if ((LA(1)=='\'') && (true)) break _loop200;
+				if ((LA(1)=='\'') && (true)) break _loop201;
 				if ((_tokenSet_8.member(LA(1))) && ((LA(2) >= '\u0000' && LA(2) <= '\ufffe'))) {
 					matchNot('\\');
 				}
@@ -1604,7 +1604,7 @@
 					mESCAPE(false);
 				}
 				else {
-					break _loop200;
+					break _loop201;
 				}
 				
 			} while (true);
@@ -1629,10 +1629,10 @@
 		_ttype = STRING2;
 		int _saveIndex;
 		
-		boolean synPredMatched203 = false;
+		boolean synPredMatched204 = false;
 		if (((LA(1)=='"') && (LA(2)=='"') && (LA(3)=='"'))) {
-			int _m203 = mark();
-			synPredMatched203 = true;
+			int _m204 = mark();
+			synPredMatched204 = true;
 			inputState.guessing++;
 			try {
 				{
@@ -1640,24 +1640,24 @@
 				}
 			}
 			catch (RecognitionException pe) {
-				synPredMatched203 = false;
+				synPredMatched204 = false;
 			}
-			rewind(_m203);
+			rewind(_m204);
 			inputState.guessing--;
 		}
-		if ( synPredMatched203 ) {
+		if ( synPredMatched204 ) {
 			_saveIndex=text.length();
 			mQUOTE3D(false);
 			text.setLength(_saveIndex);
 			{
-			_loop208:
+			_loop209:
 			do {
 				// nongreedy exit test
-				if ((LA(1)=='"') && (LA(2)=='"') && (LA(3)=='"')) break _loop208;
-				boolean synPredMatched206 = false;
+				if ((LA(1)=='"') && (LA(2)=='"') && (LA(3)=='"')) break _loop209;
+				boolean synPredMatched207 = false;
 				if (((LA(1)=='\n'||LA(1)=='\r') && ((LA(2) >= '\u0000' && LA(2) <= '\ufffe')) && ((LA(3) >= '\u0000' && LA(3) <= '\ufffe')))) {
-					int _m206 = mark();
-					synPredMatched206 = true;
+					int _m207 = mark();
+					synPredMatched207 = true;
 					inputState.guessing++;
 					try {
 						{
@@ -1665,12 +1665,12 @@
 						}
 					}
 					catch (RecognitionException pe) {
-						synPredMatched206 = false;
+						synPredMatched207 = false;
 					}
-					rewind(_m206);
+					rewind(_m207);
 					inputState.guessing--;
 				}
-				if ( synPredMatched206 ) {
+				if ( synPredMatched207 ) {
 					mNL(false);
 				}
 				else if ((_tokenSet_8.member(LA(1))) && ((LA(2) >= '\u0000' && LA(2) <= '\ufffe')) && ((LA(3) >= '\u0000' && LA(3) <= '\ufffe'))) {
@@ -1682,7 +1682,7 @@
 					mESCAPE(false);
 				}
 				else {
-					break _loop208;
+					break _loop209;
 				}
 				
 			} while (true);
@@ -1696,10 +1696,10 @@
 			match('"');
 			text.setLength(_saveIndex);
 			{
-			_loop210:
+			_loop211:
 			do {
 				// nongreedy exit test
-				if ((LA(1)=='"') && (true)) break _loop210;
+				if ((LA(1)=='"') && (true)) break _loop211;
 				if ((_tokenSet_8.member(LA(1))) && ((LA(2) >= '\u0000' && LA(2) <= '\ufffe'))) {
 					matchNot('\\');
 				}
@@ -1707,7 +1707,7 @@
 					mESCAPE(false);
 				}
 				else {
-					break _loop210;
+					break _loop211;
 				}
 				
 			} while (true);
@@ -1732,10 +1732,10 @@
 		_ttype = SEP_OR_PATH;
 		int _saveIndex;
 		
-		boolean synPredMatched138 = false;
+		boolean synPredMatched139 = false;
 		if (((LA(1)=='.') && (true) && (true))) {
-			int _m138 = mark();
-			synPredMatched138 = true;
+			int _m139 = mark();
+			synPredMatched139 = true;
 			inputState.guessing++;
 			try {
 				{
@@ -1783,12 +1783,12 @@
 				}
 			}
 			catch (RecognitionException pe) {
-				synPredMatched138 = false;
+				synPredMatched139 = false;
 			}
-			rewind(_m138);
+			rewind(_m139);
 			inputState.guessing--;
 		}
-		if ( synPredMatched138 ) {
+		if ( synPredMatched139 ) {
 			mDOT(false);
 			if ( inputState.guessing==0 ) {
 				_ttype = PATH ;
@@ -2026,7 +2026,7 @@
 		
 		match("#");
 		{
-		_loop166:
+		_loop167:
 		do {
 			if ((_tokenSet_0.member(LA(1)))) {
 				{
@@ -2034,7 +2034,7 @@
 				}
 			}
 			else {
-				break _loop166;
+				break _loop167;
 			}
 			
 		} while (true);
@@ -2062,10 +2062,10 @@
 		_ttype = NL;
 		int _saveIndex;
 		
-		boolean synPredMatched173 = false;
+		boolean synPredMatched174 = false;
 		if (((LA(1)=='\r') && (LA(2)=='\n') && (true))) {
-			int _m173 = mark();
-			synPredMatched173 = true;
+			int _m174 = mark();
+			synPredMatched174 = true;
 			inputState.guessing++;
 			try {
 				{
@@ -2073,19 +2073,19 @@
 				}
 			}
 			catch (RecognitionException pe) {
-				synPredMatched173 = false;
+				synPredMatched174 = false;
 			}
-			rewind(_m173);
+			rewind(_m174);
 			inputState.guessing--;
 		}
-		if ( synPredMatched173 ) {
+		if ( synPredMatched174 ) {
 			mNL1(false);
 		}
 		else {
-			boolean synPredMatched175 = false;
+			boolean synPredMatched176 = false;
 			if (((LA(1)=='\n'))) {
-				int _m175 = mark();
-				synPredMatched175 = true;
+				int _m176 = mark();
+				synPredMatched176 = true;
 				inputState.guessing++;
 				try {
 					{
@@ -2093,19 +2093,19 @@
 					}
 				}
 				catch (RecognitionException pe) {
-					synPredMatched175 = false;
+					synPredMatched176 = false;
 				}
-				rewind(_m175);
+				rewind(_m176);
 				inputState.guessing--;
 			}
-			if ( synPredMatched175 ) {
+			if ( synPredMatched176 ) {
 				mNL2(false);
 			}
 			else {
-				boolean synPredMatched177 = false;
+				boolean synPredMatched178 = false;
 				if (((LA(1)=='\r') && (true) && (true))) {
-					int _m177 = mark();
-					synPredMatched177 = true;
+					int _m178 = mark();
+					synPredMatched178 = true;
 					inputState.guessing++;
 					try {
 						{
@@ -2113,12 +2113,12 @@
 						}
 					}
 					catch (RecognitionException pe) {
-						synPredMatched177 = false;
+						synPredMatched178 = false;
 					}
-					rewind(_m177);
+					rewind(_m178);
 					inputState.guessing--;
 				}
-				if ( synPredMatched177 ) {
+				if ( synPredMatched178 ) {
 					mNL3(false);
 				}
 				else {
@@ -2276,10 +2276,10 @@
 		match('\\');
 		text.setLength(_saveIndex);
 		{
-		boolean synPredMatched216 = false;
+		boolean synPredMatched217 = false;
 		if (((_tokenSet_10.member(LA(1))) && ((LA(2) >= '\u0000' && LA(2) <= '\ufffe')) && (true))) {
-			int _m216 = mark();
-			synPredMatched216 = true;
+			int _m217 = mark();
+			synPredMatched217 = true;
 			inputState.guessing++;
 			try {
 				{
@@ -2287,12 +2287,12 @@
 				}
 			}
 			catch (RecognitionException pe) {
-				synPredMatched216 = false;
+				synPredMatched217 = false;
 			}
-			rewind(_m216);
+			rewind(_m217);
 			inputState.guessing--;
 		}
-		if ( synPredMatched216 ) {
+		if ( synPredMatched217 ) {
 			mESC_CHAR(false);
 		}
 		else if (((LA(1) >= '\u0000' && LA(1) <= '\ufffe')) && ((LA(2) >= '\u0000' && LA(2) <= '\ufffe')) && (true)) {
Index: src/de/fuberlin/wiwiss/ng4j/trig/parser/TriGAntlrParser.java
===================================================================
RCS file: /cvsroot/ng4j/ng4j/src/de/fuberlin/wiwiss/ng4j/trig/parser/TriGAntlrParser.java,v
retrieving revision 1.7
diff -u -r1.7 TriGAntlrParser.java
--- src/de/fuberlin/wiwiss/ng4j/trig/parser/TriGAntlrParser.java	15 Jun 2007 16:05:31 -0000	1.7
+++ src/de/fuberlin/wiwiss/ng4j/trig/parser/TriGAntlrParser.java	25 Mar 2010 15:44:32 -0000
@@ -141,6 +141,7 @@
 					break;
 				}
 				case QNAME:
+				case LCURLY:
 				case URIREF:
 				{
 					namedGraph();
@@ -195,13 +196,30 @@
 		AST namedGraph_AST = null;
 		AST label_AST = null;
 		
-		graphLabel();
-		label_AST = (AST)returnAST;
 		{
 		switch ( LA(1)) {
-		case NAME_OP:
+		case QNAME:
+		case URIREF:
 		{
-			match(NAME_OP);
+			graphLabel();
+			label_AST = (AST)returnAST;
+			{
+			switch ( LA(1)) {
+			case NAME_OP:
+			{
+				match(NAME_OP);
+				break;
+			}
+			case LCURLY:
+			{
+				break;
+			}
+			default:
+			{
+				throw new NoViableAltException(LT(1), getFilename());
+			}
+			}
+			}
 			break;
 		}
 		case LCURLY:
@@ -398,7 +416,7 @@
 		{
 			statement();
 			{
-			_loop13:
+			_loop14:
 			do {
 				if ((LA(1)==SEP)) {
 					AST tmp8_AST = null;
@@ -431,7 +449,7 @@
 					}
 				}
 				else {
-					break _loop13;
+					break _loop14;
 				}
 				
 			} while (true);
@@ -552,7 +570,7 @@
 		n_AST = (AST)returnAST;
 		astFactory.addASTChild(currentAST, returnAST);
 		{
-		_loop25:
+		_loop26:
 		do {
 			switch ( LA(1)) {
 			case PATH:
@@ -587,7 +605,7 @@
 			}
 			default:
 			{
-				break _loop25;
+				break _loop26;
 			}
 			}
 		} while (true);
@@ -1248,10 +1266,10 @@
 		AST literalModifier1_AST = null;
 		AST dt_AST = null;
 		
-		boolean synPredMatched41 = false;
+		boolean synPredMatched42 = false;
 		if (((LA(1)==AT_LANG))) {
-			int _m41 = mark();
-			synPredMatched41 = true;
+			int _m42 = mark();
+			synPredMatched42 = true;
 			inputState.guessing++;
 			try {
 				{
@@ -1259,12 +1277,12 @@
 				}
 			}
 			catch (RecognitionException pe) {
-				synPredMatched41 = false;
+				synPredMatched42 = false;
 			}
-			rewind(_m41);
+			rewind(_m42);
 			inputState.guessing--;
 		}
-		if ( synPredMatched41 ) {
+		if ( synPredMatched42 ) {
 			AST tmp29_AST = null;
 			tmp29_AST = astFactory.create(LT(1));
 			astFactory.addASTChild(currentAST, tmp29_AST);
@@ -1272,10 +1290,10 @@
 			literalModifier1_AST = (AST)currentAST.root;
 		}
 		else {
-			boolean synPredMatched43 = false;
+			boolean synPredMatched44 = false;
 			if (((LA(1)==DATATYPE))) {
-				int _m43 = mark();
-				synPredMatched43 = true;
+				int _m44 = mark();
+				synPredMatched44 = true;
 				inputState.guessing++;
 				try {
 					{
@@ -1283,12 +1301,12 @@
 					}
 				}
 				catch (RecognitionException pe) {
-					synPredMatched43 = false;
+					synPredMatched44 = false;
 				}
-				rewind(_m43);
+				rewind(_m44);
 				inputState.guessing--;
 			}
-			if ( synPredMatched43 ) {
+			if ( synPredMatched44 ) {
 				AST tmp30_AST = null;
 				tmp30_AST = astFactory.create(LT(1));
 				astFactory.addASTChild(currentAST, tmp30_AST);
Index: src/de/fuberlin/wiwiss/ng4j/trig/parser/trig.g
===================================================================
RCS file: /cvsroot/ng4j/ng4j/src/de/fuberlin/wiwiss/ng4j/trig/parser/trig.g,v
retrieving revision 1.9
diff -u -r1.9 trig.g
--- src/de/fuberlin/wiwiss/ng4j/trig/parser/trig.g	20 Feb 2009 08:09:52 -0000	1.9
+++ src/de/fuberlin/wiwiss/ng4j/trig/parser/trig.g	25 Mar 2010 15:44:32 -0000
@@ -161,7 +161,7 @@
 		;
 
 namedGraph!
-	: label:graphLabel (NAME_OP!)? graph[#label] ;
+	: (label:graphLabel (NAME_OP!)? )? graph[#label] ;
 
 graphLabel!
 	: q:qname {#graphLabel=#q;}
Index: src/de/fuberlin/wiwiss/ng4j/trig/NamedGraphSetPopulator.java
===================================================================
RCS file: /cvsroot/ng4j/ng4j/src/de/fuberlin/wiwiss/ng4j/trig/NamedGraphSetPopulator.java,v
retrieving revision 1.13
diff -u -r1.13 NamedGraphSetPopulator.java
--- src/de/fuberlin/wiwiss/ng4j/trig/NamedGraphSetPopulator.java	22 Apr 2009 17:23:50 -0000	1.13
+++ src/de/fuberlin/wiwiss/ng4j/trig/NamedGraphSetPopulator.java	25 Mar 2010 15:44:31 -0000
@@ -73,7 +73,7 @@
 	{
 		this.namedGraphSet = ngs;
 		this.base = _base ;
-		this.defaultGraphName = this.defaultGraphName == null ?
+		this.defaultGraphName = defaultGraphName == null ?
 				null : Node.createURI(defaultGraphName);
 		if ( VERBOSE )
 			System.out.println("N3toRDF: "+base) ;
