### Eclipse Workspace Patch 1.0
#P pubby-r910-patched
Index: src/de/fuberlin/wiwiss/pubby/servlets/BaseServlet.java
===================================================================
--- src/de/fuberlin/wiwiss/pubby/servlets/BaseServlet.java	(revision 910)
+++ src/de/fuberlin/wiwiss/pubby/servlets/BaseServlet.java	(working copy)
@@ -29,7 +29,7 @@
  * @version $Id$
  */
 public abstract class BaseServlet extends HttpServlet {
-	private final static String SERVER_CONFIGURATION =
+	public final static String SERVER_CONFIGURATION =
 		BaseServlet.class.getName() + ".serverConfiguration";
 	
 	private Configuration config;
@@ -85,8 +85,10 @@
 	
 	public void doGet(HttpServletRequest request,
 			HttpServletResponse response) throws IOException, ServletException {
-		String relativeURI = request.getRequestURI().substring(
-				request.getContextPath().length() + request.getServletPath().length());
+		String relativeURI = (request.getRequestURI().startsWith("/resource")) ?
+			request.getRequestURI().substring(request.getContextPath().length()) :
+			request.getRequestURI().substring(request.getContextPath().length() + request.getServletPath().length());
+			
 		// Some servlet containers keep the leading slash, some don't
 		if (!"".equals(relativeURI) && "/".equals(relativeURI.substring(0, 1))) {
 			relativeURI = relativeURI.substring(1);
Index: src/de/fuberlin/wiwiss/pubby/Configuration.java
===================================================================
--- src/de/fuberlin/wiwiss/pubby/Configuration.java	(revision 910)
+++ src/de/fuberlin/wiwiss/pubby/Configuration.java	(working copy)
@@ -158,4 +158,8 @@
 	public String getWebApplicationBaseURI() {
 		return config.getProperty(CONF.webBase).getResource().getURI();
 	}
+	
+	public Collection getDatasets() {
+		return datasets;
+	}
 }
